### ISSUE TITLE
go--directory-dirs does not deal well with "circular" symlinks?

### ISSUE THREAD

**Issue by ccressent** [dominikh/go-mode.el#278]
I'm working on a project that has a dependency on `etcd` v.3.3.2, which has ["circular" symlinks in it](https://github.com/etcd-io/etcd/blob/v3.3.2/cmd/etcd) ([other example](https://github.com/etcd-io/etcd/blob/v3.3.2/cmd/vendor/github.com/coreos/etcd)). Whenever I call `go-guru-callers` in this project it inevitably leads to a `Variable binding depth exceeds max-specpdl-size` error, no matter what the value of `max-specpdl-size` is.

My guess is that the `go--directory-dirs` function does not deal well with those kinds of "circular" links? See Emacs debugger trace below.

<details>
<summary>Full debugger trace</summary>

```
Debugger entered--Lisp error: (error "Variable binding depth exceeds max-specpdl-size")
  file-directory-p("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-co...")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com/coreos")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor/github.com")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd/vendor")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd/cmd")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos/etcd")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com/coreos")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor/github.com")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd/vendor")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd/cmd")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources/git---github.com-coreos-etcd")
  go--directory-dirs("/home/cyril/go/pkg/dep/sources")
  go--directory-dirs("/home/cyril/go/pkg/dep")
  go--directory-dirs("/home/cyril/go/pkg/")
  #f(compiled-function (topdir) #<bytecode 0x158c2bcee42d>)("/home/cyril/go")
  mapcan(#f(compiled-function (topdir) #<bytecode 0x158c2bcee42d>) ("/usr/lib/go" "/home/cyril/go"))
  cl-mapcan(#f(compiled-function (topdir) #<bytecode 0x158c2bcee42d>) ("/usr/lib/go" "/home/cyril/go"))
  go-packages-native()
  go-packages()
  go-guru--read-scope()
  go-guru-set-scope()
  go-guru--set-scope-if-empty()
  go-guru-callers()
  funcall-interactively(go-guru-callers)
  call-interactively(go-guru-callers nil nil)
  command-execute(go-guru-callers)
```
</details>

**Comment by dominikh** [comment]
You're correct. The underlying issue is that `file-directory-p` returns true for symlinks to directories. Will fix.

**Comment by dominikh** [comment]
Should be fixed by 287595a3

**Comment by ccressent** [comment]
Thank you @dominikh!
Awesome explanation in the commit message too!
